fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/typing/Type.java

```diff
@@ -1259,6 +1259,99 @@ static Any any() {
         return Any.INSTANCE;
     }
 
+    /**
+     * Special {@link Type.Record} that is undefined.
+     */
+    class AnyRecord extends Record {
+        @Nonnull
+        private final Supplier<Integer> hashCodeSupplier = Suppliers.memoize(this::computeHashCode);
+
+        @Nonnull
+        private static final AnyRecord INSTANCE = new AnyRecord();
+
+        private int computeHashCode() {
+            return Objects.hash(getTypeCode().name().hashCode(), isNullable());
+        }
+
+        public AnyRecord() {
+            super(true, List.of());
+        }
+
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public boolean isNullable() {
+            return true;
+        }
+
+        @Nonnull
+        @Override
+        public AnyRecord withNullability(final boolean newIsNullable) {
+            Verify.verify(newIsNullable);
+            return this;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public void addProtoField(@Nonnull final TypeRepository.Builder typeRepositoryBuilder,
+                                  @Nonnull final DescriptorProto.Builder descriptorBuilder,
+                                  final int fieldNumber,
+                                  @Nonnull final String fieldName,
+                                  @Nonnull final Optional<String> typeNameOptional,
+                                  @Nonnull final FieldDescriptorProto.Label label) {
+            throw new UnsupportedOperationException("type any cannot be represented in protobuf");
+        }
+
+        @Override
+        public int hashCode() {
+            return hashCodeSupplier.get();
+        }
+
+        @Override
+        public boolean equals(final Object obj) {
+            if (obj == null) {
+                return false;
+            }
+
+            if (this == obj) {
+                return true;
+            }
+
+            if (getClass() != obj.getClass()) {
+                return false;
+            }
+
+            final var otherType = (Type)obj;
+            return getTypeCode() == otherType.getTypeCode() && isNullable() == otherType.isNullable();
+        }
+
+        @Override
+        public String toString() {
+            return getTypeCode().toString();
+        }
+
+        @Nonnull
+        @Override
+        public PRecordType toProto(@Nonnull final PlanSerializationContext serializationContext) {
+            return PRecordType.newBuilder().build();
+        }
+
+        @Nonnull
+        @Override
+        public RecordQueryPlanProto.PType toTypeProto(@Nonnull final PlanSerializationContext serializationContext) {
```

Comment:
Specifically,
- add `PAnyRecord` into `record_query_plan.proto`
- add a `Deserializer` as a nested class using the `@AutoService` (see other implementors in `Type`)
- change `toProto()` above to create a `PAnyRecord` instead of an empty `PRecordType`
- add a test